File: queue/Q641.java
public class Q641 {
    private static class MyCircularDeque {
        private final int[] values;
        private int size;
        private int front;
        private int end;

        /** Initialize your data structure here. Set the size of the deque to be k. */
        public MyCircularDeque(int k) {
            values = new int[k];
        }

        private boolean init(int value) {
            if (!isEmpty()) return false;

            front = 0;
            end = 0;
            values[0] = value;
            size = 1;
            return true;
        }

        /** Adds an item at the front of Deque. Return true if the operation is successful. */
        public boolean insertFront(int value) {
            if (isFull()) return false;
            if (isEmpty()) return init(value);

            int newFront = Math.floorMod(front - 1, values.length);
            values[newFront] = value;
            front = newFront;
            size++;
            return true;
        }

        /** Adds an item at the rear of Deque. Return true if the operation is successful. */
        public boolean insertLast(int value) {
            if (isFull()) return false;
            if (isEmpty()) return init(value);

            int newEnd = Math.floorMod(end + 1, values.length);
            values[newEnd] = value;
            end = newEnd;
            size++;
            return true;
        }

        /** Deletes an item from the front of Deque. Return true if the operation is successful. */
        public boolean deleteFront() {
            if (isEmpty()) return false;

            front = Math.floorMod(front + 1, values.length);
            size--;
            return true;
        }

        /** Deletes an item from the rear of Deque. Return true if the operation is successful. */
        public boolean deleteLast() {
            if (isEmpty()) return false;

            end = Math.floorMod(end - 1, values.length);
            size--;
            return true;
        }

        /** Get the front item from the deque. */
        public int getFront() {
            if (isEmpty()) return -1;

            return values[front];
        }

        /** Get the last item from the deque. */
        public int getRear() {
            if (isEmpty()) return -1;

            return values[end];
        }

        /** Checks whether the circular deque is empty or not. */
        public boolean isEmpty() {
            return size == 0;
        }

        /** Checks whether the circular deque is full or not. */
        public boolean isFull() {
            return size == values.length;
        }
    }
}

File: LinkedList-Stack-Queue/Q641.py
# https://leetcode.com/problems/design-circular-deque/


class MyCircularDeque:

    def __init__(self, k: int):
        """
        Initialize your data structure here. Set the size of the deque to be k.
        """
        self.deque = [-1] * k
        self.size = 0
        self.head = 0
        self.tail = 0

    def insertFront(self, value: int) -> bool:
        """
        Adds an item at the front of Deque. Return true if the operation is successful.
        """

        if self.isFull():
            return False

        if not self.isEmpty():
            self.head = (self.head - 1) % len(self.deque)
        self.deque[self.head] = value
        self.size += 1
        return True

    def insertLast(self, value: int) -> bool:
        """
        Adds an item at the rear of Deque. Return true if the operation is successful.
        """
        if self.isFull():
            return False

        if not self.isEmpty():
            self.tail = (self.tail + 1) % len(self.deque)
        self.deque[self.tail] = value
        self.size += 1
        return True

    def deleteFront(self) -> bool:
        """
        Deletes an item from the front of Deque. Return true if the operation is successful.
        """
        if self.isEmpty():
            return False

        self.head = (self.head + 1) % len(self.deque)
        self.size -= 1
        if self.size <= 0:
            self.head = 0
            self.tail = 0
        return True

    def deleteLast(self) -> bool:
        """
        Deletes an item from the rear of Deque. Return true if the operation is successful.
        """
        if self.isEmpty():
            return False

        self.tail = (self.tail - 1) % len(self.deque)
        self.size -= 1
        if self.size <= 0:
            self.head = 0
            self.tail = 0
        return True

    def getFront(self) -> int:
        """
        Get the front item from the deque.
        """
        return self.deque[self.head] if not self.isEmpty() else -1

    def getRear(self) -> int:
        """
        Get the last item from the deque.
        """
        return self.deque[self.tail] if not self.isEmpty() else -1

    def isEmpty(self) -> bool:
        """
        Checks whether the circular deque is empty or not.
        """
        return self.size <= 0

    def isFull(self) -> bool:
        """
        Checks whether the circular deque is full or not.
        """
        return self.size == len(self.deque)

# Your MyCircularDeque object will be instantiated and called as such:
# obj = MyCircularDeque(k)
# param_1 = obj.insertFront(value)
# param_2 = obj.insertLast(value)
# param_3 = obj.deleteFront()
# param_4 = obj.deleteLast()
# param_5 = obj.getFront()
# param_6 = obj.getRear()
# param_7 = obj.isEmpty()
# param_8 = obj.isFull()


["MyCircularDeque", "insertLast", "insertLast", "insertFront", "insertFront",
    "getRear", "isFull", "deleteLast", "insertFront", "getFront"]
[[3], [1], [2], [3], [4], [], [], [], [4], []]
